tket/src/Transformations/include/Transformations/PhasedXFrontier.hpp

```diff
@@ -0,0 +1,274 @@
+ * single-qubit subcircuit of `circ` forming a sequence of single-qubit gates
+ * between two multi-qb gates (or the input/output vertices) of the original
+ * circuit. Each interval is defined by start and end edges, which are edges of
+ * the circuit DAG. end is in the future of start. The source of the start edge
+ * is a multi-qb gate (or an Input vertex) and the target of the end edge is
+ * a multi-qb gate (or an Output vertex).
+ *
+ * ## Initialisation and moving forward
+ * The frontier is initialised at the beginning of the circuit, at which stage
+ * an interval is created for each qubit of `circ`.
+ * The frontier then moves forward in jumps, where the new start of the interval
+ * will be the edge after the former end edge. The new end edge will be moved
+ * forward along the qubit until its target is a multi-qb gate or the Output
+ * vertex.
+ *
+ * One can move forward an individual interval by calling `next_interval`.
+ * Careful, however: all intervals are assumed to be space-like separated (i.e.
+ * there cannot be a path between an interval and another at any time).
+ * Moving arbitrary intervals forward may break this invariant.
+ * To avoid this issue, the `next_multiqb` method is provided. It takes a multi-
+ * qb vertex `v` as argument and will move forward all the qubits whose end edge
+ * has target `v`. By calling `next_multiqb` on each multi-qb gate in a causal
+ * order, the space-like separation invariant will be maintained.
+ *
+ * ## Operations on intervals
+ * ### Squashing
+ * Each interval can be squashed into a normal form `PhasedX + Rz` using
+ * `squash_interval`, or `squash_intervals` to squash all current intervals.
+ * It is recommended to do this before acting on the intervals, as otherwise
+ * multiple single-qb operations can accumulate and lead to inefficient
+ * decompositions. This is what `globalise_PhasedX(squash=true)` does
+ * (the default). This is however not necessary for the frontier to be well-
+ * behaved
+ *
+ * ### Get beta angles
+ * When squashed, each interval is associated with a unique "beta angle", given
+ * by the formula U_2 = Rz(α)Rx(β)Rz(ɣ) = PhasedX(β, α)Rz(ɣ - α).
+ * These can be obtained for each interval using `get_all_betas`, while
+ * `get_beta_edge` and `get_beta_edges` will return the edge pointing to the
+ * gate that contains the β-angle (ie the PhasedX gate).
+ *
+ * ### Insert global NPhasedX gates
+ * `insert_1_phasedx` and `insert_2_phasedx` will insert 1x resp 2x global
+ * NPhasedX gates at the beginning of the current frontier. The resuling
+ * circuit will always be equivalent to the original one.
+ */
+class PhasedXFrontier {
+ public:
+  /**
+   * @brief Construct a new Phased X Frontier object.
+   *
+   * @param circ The circuit to traverse.
+   */
+  PhasedXFrontier(Circuit& circ);
+
+  /**
+   * @brief Squash the current intervals on each qubit
+   */
+  void squash_intervals();
```

Comment:
It would be easy to add, but in this case it would not be useful. The main issue with detecting changes in the circuit is the decomposition and resynthesis of global gates, see discussion above.